### PR TITLE
chore: get rid of `Name` and refactor

### PR DESCRIPTION
**Pull request by harrysolovay**
This PR removes the `Name` type, which was previously used to hide exorbitantly-long signatures with a pretty, effect-specific name. Since we now flatten all signatures into a single `Effect` type, this is no longer necessary.

Some other minor changes:

- Renaming `core` dir to `effects` (we can reorganize further as we create more)
- `as` method attached to `Effect` for producing same effect with narrowed `T`
- `access` method attached to `Effect` for producing new effect to resolve specified index of result

### CODE REVIEW

File: Effect.ts [hunk 1]
```diff
@@ -24,36 +24,34 @@ export class Effect<
...
+      return U.memo(() => {
+        return U.thenOk(
+          U.all(process.resolve(this), process.resolve(key)),
+          ([target, key]) => target[key],
+        );
+      });
+    }, [this, key]);
+  };
+
+  as = <Satisfies extends T>() => {
```
Comment (tjjfvi) [inline Effect.ts:41]: Why `Satisfies`?

Comment (harrysolovay) [reply Effect.ts:41]: It's for narrowing, so the supplied type should ideally extend the pre-existing `T`, no? The word "satisfies" is also a gentle reminder for us to (possibly) eventually make use of TS 4.9's `satisfies` operator.

Comment (tjjfvi) [reply Effect.ts:41]: `satisfies` is for safe upcasts, but this is an unsafe downcast. It should really be `U extends T` IMO.

Comment (harrysolovay) [reply Effect.ts]: Fair enough, just updated

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**